uint256/uint256.go

```diff
@@ -0,0 +1,142 @@
+	return u
+}
+
+// Value returns the underlying private value for a Uint256 struct
+func (u *Uint256) Value() big.Int {
+	return u.value
+}
+
+// Set assigns the underlying value of the given Uint256 param to u, and returns the modified receiver struct
+// returns an error when the result falls outside of the unsigned 256-bit integer range
+func (u *Uint256) Set(value big.Int) (*Uint256, error) {
+	if value.Cmp(maxUint256) == 1 {
+		return nil, fmt.Errorf("cannot set Uint256 to %v as it overflows max value of %v", value, maxUint256)
+	}
+	if value.Cmp(minUint256) == -1 {
+		return nil, fmt.Errorf("cannot set Uint256 to %v as it underflows min value of %v", value, minUint256)
+	}
+	u.value = *new(big.Int).Set(&value)
+	return u, nil
+}
+
+// Copy sets the underlying value of u to a copy of the given Uint256 param, and returns the modified receiver struct
+func (u *Uint256) Copy(v *Uint256) *Uint256 {
+	u.value = *new(big.Int).Set(&v.value)
+	return u
+}
+
+// Cmp calls the underlying Cmp method for the big.Int stored in a Uint256 struct as its value field
+func (u *Uint256) Cmp(v *Uint256) int {
+	return u.value.Cmp(&v.value)
+}
+
+// Equals returns true if the two Uint256 structs have the same underlying values, false otherwise
+func (u *Uint256) Equals(v *Uint256) bool {
+	return u.Cmp(v) == 0
+}
+
+// Add sets u to augend + addend and returns u as the sum
+// returns an error when the result falls outside of the unsigned 256-bit integer range
+func (u *Uint256) Add(augend, addend *Uint256) (*Uint256, error) {
+	sum := new(big.Int).Add(&augend.value, &addend.value)
+	return u.Set(*sum)
+}
+
+// Sub sets u to minuend - subtrahend and returns u as the difference
+// returns an error when the result falls outside of the unsigned 256-bit integer range
+func (u *Uint256) Sub(minuend, subtrahend *Uint256) (*Uint256, error) {
+	difference := new(big.Int).Sub(&minuend.value, &subtrahend.value)
+	return u.Set(*difference)
+}
+
+// Mul sets u to multiplicand * multiplier and returns u as the product
+// returns an error when the result falls outside of the unsigned 256-bit integer range
+func (u *Uint256) Mul(multiplicand, multiplier *Uint256) (*Uint256, error) {
+	product := new(big.Int).Mul(&multiplicand.value, &multiplier.value)
+	return u.Set(*product)
+}
+
+// String returns the string representation for Uint256 structs
+func (u *Uint256) String() string {
+	return u.value.String()
+}
+
+// MarshalJSON implements the json.Marshaler interface
+// it specifies how to marshal a Uint256 struct so that it can be written to disk
+func (u Uint256) MarshalJSON() ([]byte, error) {
+	return []byte(u.value.String()), nil
+}
+
+// UnmarshalJSON implements the json.Unmarshaler interface
+// it specifies how to unmarshal a Uint256 struct so that it can be reconstructed from disk
+func (u *Uint256) UnmarshalJSON(b []byte) error {
+	if string(b) == "null" {
+		return nil
+	}
+
+	var value big.Int
+	_, ok := value.SetString(string(b), 10)
```

Comment:
Shouldn't we check the range at this point as well?

Reply:
do you mean the [0, 2^256 - 1] range?

if so: we don't normally expect to marshal or unmarshal a `Uint256` variable which has an invalid `value`, right? 

if they are created and updated correctly, then it follows that only correct values can be marshaled/unmarshaled. but of course we don't have this certainty.

so, do you think we should take this extra step in case something happens between marshaling a valid `Uint256` and unmarshaling it?

Reply:
Well unmarshalling is when you get a value from JSON, and that is either through some remote call or some deserialization from file or something like that - which implicitly means it is outside of our direct control, or in other words, you don't know where that value *could* come from, that is why to be really safe the range should be checked IMO.

Reply:
you're absolutely right. on it

Reply:
you're absolutely right. done